barretenberg/cpp/src/barretenberg/vm/avm/trace/gadgets/merkle_tree.cpp

```diff
@@ -0,0 +1,103 @@
+#include "barretenberg/vm/avm/trace/gadgets/merkle_tree.hpp"
+
+namespace bb::avm_trace {
+
+AvmMerkleTreeTraceBuilder::MerkleEntry AvmMerkleTreeTraceBuilder::compute_root_from_path(uint32_t clk,
+                                                                                         const FF leaf_value,
+                                                                                         const uint32_t leaf_index,
+                                                                                         const std::vector<FF>& path)
+{
+    uint32_t path_length = static_cast<uint32_t>(path.size());
+    FF curr_value = leaf_value;
+    uint32_t curr_index = leaf_index;
+    std::vector<bool> path_bits;
+    std::vector<FF> path_values;
+    // These will be eventually stored somewhere as a "clock speed"
+    auto entry_id = clk << 6;
+    for (uint32_t i = 0; i < path_length; i++) {
+        bool path_parity = (curr_index % 2 == 0);
+
+        curr_value =
+            path_parity
+                ? poseidon2_builder.poseidon2_hash({ curr_value, path[i] }, entry_id + i, Poseidon2Caller::MERKLE_TREE)
+                : poseidon2_builder.poseidon2_hash({ path[i], curr_value }, entry_id + i, Poseidon2Caller::MERKLE_TREE);
+        path_values.push_back(curr_value);
+        curr_index >>= 1;
+    }
+    return MerkleEntry{ .clk = clk,
+                        .leaf_value = leaf_value,
+                        .leaf_index = leaf_index,
+                        .path = path,
+                        .path_bits = path_bits,
+                        .path_values = path_values,
+                        .root = curr_value };
+}
+
+bool AvmMerkleTreeTraceBuilder::check_membership(
+    uint32_t clk, const FF leaf_value, const uint32_t leaf_index, const std::vector<FF>& path, const FF& root)
+{
+    MerkleEntry entry = compute_root_from_path(clk, leaf_value, leaf_index, path);
+    entry.is_membership_op = true;
+    // If the computed root is the same as the expected then the leaf is a member
+    bool is_member = entry.root == root;
+    entry.is_member = is_member;
+    // If the leaf is not a member then we should replace the computed root with the expected root
+    entry.root = is_member ? entry.root : root;
+    merkle_check_trace.push_back(entry);
+    return entry.is_member;
```

Comment:
Yeah always against the current root. The merkle gadget works is that if we are "claiming membership", the circuit will constrain that the `circuit_computed_root == current_root`. While in non-membership this check becomes that `circuit_computed_root != current_root`. 

The function that calls the merkle gadget will have to individually constrain them what the non-membership checks "mean" - basically as you described.

Reply:
I'm still a bit confused and just want to make sure we're on the same page. I meant that I'm not convinced we ever want to use this gadget with `circuit_computed_root != current_root`. There isn't much value in proving that a sibling path doesn't hash to the current root, because the sequencer could provide a phony path to get the AVM to incorrectly prove non-membership.

So for non membership, I think we still always want to prove `circuit_computed_root == current_root`, and we just want to change for example which leaf we are proving exists.

Reply:
yeah i think youre right - we should just prove membership checks in the avm.